ls1intum/Artemis
# `General`: Redesign the alert system

### Checklist
#### General


#### Client

### Motivation and Context



Currently, general alerts provided by the AlertsService are inserted somewhere in the page itself. This causes movements of other, relevant content when alerts appear and disappear, which can cause disturbances and irritation.
Additionally, alerts may not be visible if you scroll away from the alert mount location.

Alerts should therefore be floating above other content to not randomly move around other page content, should be visible regardless to scrolling position, and be animated to raise the users attention if they come in.

Closes #3148

### Description


This PR removes ``<jhi-alert>`` from everywhere where it's used, and displays alerts in a global overlay that is present above all other views. It has a fixed position to be always visible when scrolling the page.

Alerts fly in from the top right if they come in, and fly back to the right if they are closed, either by timeout, another imperative call on the alert from elsewhere, or by the user dismissing the alert. If multiple alerts are present, they are display in a vertical column. Remainig alerts move their position smoothly if other alerts are added or removed.
Additionally, alerts will indicate the time until they vanish by a small bar on the bottom.
Move Animations are done using Angular Animations; the icon and the timing bar is just CSS.

New: Alerts will now accept an ``action`` when created. An ``action`` in this context contains a ``label`` and a ``callback``. Setting it will render a button with the label on the alert. If the user clicks the button, the given ``callback`` function will be called. This is optional, but can be useful to use alerts as calls to action: Alerts usually raise the attention as they move and blink when coming in, and users can immediately respond to your CTA by clicking the button if needed. The first usecase for this will be the oudated version alert where we will render an "Update now" button to upgrade the client to the new version if available.

Additionally, the alerts service has been reworked:
- Removed support for external alerts arrays
- Removed the dedicated error alert component
- Alerts can be closed by just calling a function on alert objects
- Alerts have a property ``isOpen`` that will be updated accordingly if closed
- Alerts have a ``onClose`` callback that can be set when creating the alert

Moreoever, a few other features have been altered to centralize alerts into one single service:
- Moved HTTP error handlers from the error alert component to the alerts service
- Converted the connection notification component from a component to a service that just adds websocket connection state notifications to the new system

### Steps for Testing


I have added a "Add alert" link to the right side of the footer if you want to spawn a lot of testing alerts in this release:
https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=486309900 (Commit 73a4f3b)
It's not part of the PR anymore and will not be merged; it's just for testing.

Additionally, this affects all places where the alertService is used. These are a lot.
However, the PR is designed to serve as a drop in replacement and the current API of the alerts service is only extended, and no breaking changes were made. Therefore, alerts should continue to work normally everywhere where the service is used.

Please navigate through the app and try to trigger some alerts where you know they would usually appear, and check if they appar as expected, with the correct type and content, and disappear after the correct timeout.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage




- alert-overlay.component.ts: 100% | 100%
- alert.service.ts: 82.35% | 94.56%

### Screenshots




https://user-images.githubusercontent.com/23171488/151016128-31a3c1c0-1230-45f5-813a-0982790b33e0.mov
